[4.0] Menus breadcrumb backend

VladikB opened the pull request:
Redo of Issue #21361.

### Summary of Changes

Added Breadcrumb to menus in backend

### Testing Instructions

1. Go in the backend to Menus -> All Menu Items -> New
2. Call Menu "TEST" and select any Menu Item Type
3. Save and Close it
4. Add a New Menu, call it "TEST 2"  and choose any Menu Type, AND choose "TEST" as parent item

### Expected result


![Breacrumpbs](https://user-images.githubusercontent.com/45699656/64422528-94bae100-d0a4-11e9-9c5c-2ec841b741d0.PNG)

VladikB committed:
Add backend Breadcrumbs

VladikB committed:
Code cleanup

VladikB committed:
Code style

mbabker commented:
1) Views really shouldn’t be running database queries on their own accord

2) This is introducing a lot of database queries to this view

hans2103 commented:
# Before 
<img width="412" alt="Schermafdruk 2019-09-06 16 48 57" src="https://user-images.githubusercontent.com/639822/64437976-8e8a2c00-d0c7-11e9-8306-8852a19414b0.png">

# After 
<img width="445" alt="Schermafdruk 2019-09-06 16 57 54" src="https://user-images.githubusercontent.com/639822/64437977-8e8a2c00-d0c7-11e9-9993-94641fd60464.png">


Test successful, but not useful to me.
The word "test" is short... lot of my clients have longer Menu Item titles. 
Some customers even have sub sub menu items 
This will result in a very long breadcrumb in the backend. 

Please do not merge this PR into J4.... or give me the possibility to disable this feature. It might be useful to others. But not for me.

roland-d commented:
Thanks @mbabker we will look into this.

roland-d commented:
@hans2103 Can you propose a better way of displaying the breadcrumbs?

mbabker commented:
> Can you propose a better way of displaying the breadcrumbs?

Let's take it a step further.  What is the need to display a full breadcrumb to every menu item on the list view and what kind of value is it adding by doing so?  How is this an issue only for menu items and not other hierarchical data types (i.e. why not the full path for categories or tags)?

Both this issue and the original PR are very vague on why this is a needed improvement or even a good idea in general.  They both right now are written in a "I think this way is better and you should accept it" way without really explaining why this way is better.

hans2103 commented:
@roland-d

There is no need for me to display the breadcrumbs in Joomla Administrator.
It takes to much space on a mobile device. 
<img width="360" alt="Schermafdruk 2019-09-06 17 16 37" src="https://user-images.githubusercontent.com/639822/64439377-1a9d5300-d0ca-11e9-89e0-bc38d5878bf5.png">

Why propose a better way if there is no need for breadcrumbs to display on every menu item?
A better way for me is not to display the breadcrumbs. :-) (or have the option to disable it)

HLeithner commented:
@VladikB for you first contribution to joomla, 

Beside my personal opinion that the menu tree per menu item overloads the view you should use the Joomla API instead of directly querying the database. In your case you can use the menu framework to get the menu items instead calling your searchParentItem function.

Something like this should work: ``` Factory::getApplication()->getMenu()->getItem($itemid); ```

On the other side you are creating the tree in the layout file, you shouldn't do this, you should do any preparation in the model.

I would change the model getItems function and add the tree to the object.

But as already mention I don't think that adding the tree brings any benefit to the view.

roland-d commented:
Let's get back to why this even became an idea that should be fixed.

>  What is the need to display a full breadcrumb to every menu item on the list view and what kind of value is it adding by doing so? 

Here is my menu, I have no clue which menu item belongs to which parent:
<img width="609" alt="image" src="https://user-images.githubusercontent.com/359377/64452821-70ccbf00-d0e7-11e9-8959-c4685a23b5e9.png">

Showing me the parent, will give me an idea of which menu item I need to pick to change.

>  How is this an issue only for menu items and not other hierarchical data types (i.e. why not the full path for categories or tags)?

I would argue that hierarchical data types in the Joomla backend are horrible to work with. So it is not only menus but it could use a better solution in many more places. Here is another example in editing a menu item:
<img width="250" alt="image" src="https://user-images.githubusercontent.com/359377/64452927-af627980-d0e7-11e9-833b-12e58589eada.png">

Same goes for the category selector in the search tools:
<img width="303" alt="image" src="https://user-images.githubusercontent.com/359377/64452994-d9b43700-d0e7-11e9-9d9f-4f626adcd671.png">

Unless it is just me and everybody else thinks this is a normal way of working. Does this give a better insight as to the problem it tries to solve?

> It takes to much space on a mobile device.
My other idea was we show the breadcrumbs the same way the selected item is shown below the menu option.

brianteeman commented:
Definitely an issue for me with the filters

brianteeman commented:
Oh and I forgot it also solved a big accessibility issue

hans2103 commented:
@roland-d having those arguments it is a legit solution for a legit issue.
But... as you say it solves an issue when you apply filters.
So... the breadcrumbs should appear when you apply filters.
Only when you apply filters.

brianteeman commented:
The accessibility issue applied all the time

roland-d commented:
@hans2103 
> But... as you say it solves an issue when you apply filters.

Not only when you apply filters, also happens when your menu runs across multiple pages.

@brianteeman I am not fully understanding, adding these breadcrumbs solves an accessibility issue?

brianteeman commented:
Yes it solves an issue. If you can only hear one line at a time it's not possible to get the context of a menu item. Showing the full path solves that.

mbabker commented:
Having the full path in the filters is a must fix and something I've groaned about on other products too (still not fixed there either last I looked).  But that's not the focus here.

To me, the breadcrumb is really only _needed_ on the main list view when:

- You've paginated and the parent item isn't on the same page
- You've filtered and the parent item isn't anywhere to be found

To me it feels like there really should be a better way of displaying the tree information without repeating the parent labels on every line.  Especially as it feels like to me it would turn into an information overload once you start getting 2 or 3 levels deep.

![Screen Shot 2019-09-06 at 2 41 32 PM](https://user-images.githubusercontent.com/368545/64455919-82e13a00-d0b4-11e9-934f-ce677c7eae94.png)

brianteeman commented:
I agree with that as well. It's why I haven't submitted an a11y pr for it as I couldn't work out a nice look

roland-d commented:
I am no UX designer either. I guess if we move the path 1 line down, you have 3 lines of text which is a bit much as well.

mbabker commented:
Do you really need both the menu item type breadcrumb and the menu item path breadcrumb?  Surely the menu item path is going to be more useful than the item type and in a game of picking and choosing I'd rather that in a similar display as what's used in the articles list than a (IMO) superfluous path that doesn't really give me much beneficial information on that screen.

roland-d commented:
That is an interesting thought. I do agree I barely ever look at the menu item type as the title of the menu is usually more descriptive to me. So replacing one with the other would make sense to me as well.

coolcat-creations commented:
Actually the menu item type is important for beginners to understand what they are doing. Please do not remove that or at least add a symbol displaying if its an Article / Blog / List / Contact / Komponent... 
I can imagine having the breadcrumbs in small size above the title, but with the possibility to turn off the information?

VladikB committed:
Refactor loading the items
Move the path below the menu name

VladikB commented:
So it is how it looks now. I moved the breadcrumbs to a line under the title and changed the font type to a smaller size.

![Breadcrumbs1](https://user-images.githubusercontent.com/45699656/64522721-ec528a00-d2fa-11e9-809b-e7798f692b8e.PNG)

roland-d commented:
@VladikB Posted the path below the name just so we can see how it looks like.

coolcat-creations commented:
That looks quite cluttered, maybe the Alias should be below too to have a better layout of the Informations.
And like that:
Alias:                   ....
Path:                   ....
Menuitem type:  ....

hans2103 commented:
> That looks quite cluttered, maybe the Alias should be below too to have a better layout of the Informations.
> And like that:
> Alias: ....
> Path: ....
> Menuitem type: ....

Which will result in four rows per menu item:

```
Title
alias: ...
path: ...
menu item type: ...
```

Please replace `/` by `›` as the latter indicates a beter forward motion then a forward slash.

`›` is a Single right-pointing angle quotation mark
- unicode = U+0203A
- hex code = &#x203A;
- html code = &#8250;
- html entity = &rsaquo;
- css code = \203A

coolcat-creations commented:
> Which will result in four rows per menu item:
> 
> ```
> Title
> alias: ...
> path: ...
> menu item type: ...
> ```

Yes but if you look at the screenshot its probably better to have the four rows but not a cut off alias... we could give the Menu item type an own column?

VladikB committed:
Replaced / by ›

VladikB commented:
> > Which will result in four rows per menu item:
> > ```
> > Title
> > alias: ...
> > path: ...
> > menu item type: ...
> > ```
> 
> Yes but if you look at the screenshot its probably better to have the four rows but not a cut off alias... we could give the Menu item type an own column?

The cut off alias would be there without my changes too.

VladikB commented:
But it would look like that

![Breadcrumbs2](https://user-images.githubusercontent.com/45699656/64525998-76eab780-d302-11e9-9a24-0dbf2067f86e.PNG)

infograf768 commented:
Please also test this in RTL. ;)

coolcat-creations commented:
Can you remove this part?
<img width="73" alt="grafik" src="https://user-images.githubusercontent.com/828371/64526841-a8648280-d304-11e9-853c-3e5672d82ea9.png">
 Just the symbols, not the intention

coolcat-creations commented:
What abount using "ellipse" when it's getting to long and show the full line in a title on hover?
good or bad idea?

hans2103 commented:
@coolcat-creations when is a breadcrumb too long? 
On a mobile device (iPhone 5) the viewport is only 320px

coolcat-creations commented:
Maybe the breadcrumbs should be more like a title line above ALL items that are below a parent instead of repeating itself over and over again. This would need some thinking how to code it when items are torn apart in the middle.

brianteeman commented:
That would fail accessibility

coolcat-creations commented:
Ok, I don't see a nice solution, maybe the design team has an idea...

VladikB commented:
I think, it's OK how it looks like now, I mean if it was accepted that Alias can break the line like they do, why shouldn't it be aceepted with breadcrumb?

coolcat-creations commented:
I was only thinking about improving the layout for Joomla 4 - now that we have a chance to do so...

mbabker commented:
Keep alias on the same line as the title.  The same way it is done in every
other list screen.  If we want to move it then do it in a separate PR
changing all views at once.

On Mon, Sep 9, 2019 at 7:06 AM VladikB <notifications@github.com> wrote:

> I think, it's OK how it looks like now, I mean if it was accepted that
> Alias can break the line like they do, why shouldn't it be aceepted with
> breadcrumb?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/26187?email_source=notifications&email_token=AACZ7IOKI55ZSVXQU5ZE33LQIY35XA5CNFSM4IUHVBTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6HKC5I#issuecomment-529441141>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACZ7IOLD3YDOMQ3AISSIYTQIY35XANCNFSM4IUHVBTA>
> .
>
-- 
- Michael Please pardon any errors, this message was sent from my iPhone.

VladikB commented:
They are in the same line, I just showed how it would look like if they weren't

mbabker commented:
I know, but with it being suggested, I wanted to squash it early because it
should be changed in all views if that is going to be seriously pursued
(don’t care either way where it ends up), and that goes far beyond the goal
of this PR.

On Mon, Sep 9, 2019 at 7:18 AM VladikB <notifications@github.com> wrote:

> They are in the same line, I just showed how it would look like if they
> weren't
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/26187?email_source=notifications&email_token=AACZ7IIRGZ4XFXXYTEF53ADQIY5JDA5CNFSM4IUHVBTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6HLEGA#issuecomment-529445400>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACZ7IKV3X2X6SZWETKDHALQIY5JDANCNFSM4IUHVBTA>
> .
>
-- 
- Michael Please pardon any errors, this message was sent from my iPhone.

VladikB committed:
Added an option of turning on the breadcrumbs

VladikB commented:
So I added an option to turn on or to turn off the breadcrumbs now.

VladikB commented:
![Option_Breadcrumbs](https://user-images.githubusercontent.com/45699656/64684182-edafbe00-d484-11e9-9f75-e9cff9086700.PNG)

coolcat-creations commented:
Shouldn't that be a global option somewhere else? For me "Page Display" options are for the Frontend... We have column display options in the tables asfaik

ahghatol commented:
Failed - tested on d9fbed64bf

priyankaSahutekdi commented:
I have tested this item :red_circle: unsuccessfully on d9fbed64bf3815d877f4ab759d28f966ffd00852<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26187">issues.joomla.org/tracker/joomla-cms/26187</a>.</sub>

Renuka-S commented:
I have tested this item :white_check_mark: successfully on d9fbed64bf3815d877f4ab759d28f966ffd00852<br /><br />If the final output is expected as below then the test is successful. 
 
If Menu Options > Show breadcrumbs is "ON" then showing the result as
- Menu Title Alias:
- Breadcrumb
- Menu Item type

Else if Menu Options > Show breadcrumbs is "OFF" then showing the result as
- Menu Title Alias:
- Menu Item type

Please, can someone confirm what is the expected final output? Hard to understand from comments.
 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26187">issues.joomla.org/tracker/joomla-cms/26187</a>.</sub>

Renuka-S commented:
Adding the screenshots for above successful![screen shot 2019-10-19 at 09 50 21](https://issues.joomla.org/uploads/1/d66a2cbbe45e1caa72d2ee4eb2fb180e.png) test![screen shot 2019-10-19 at 09 50 23](https://issues.joomla.org/uploads/1/6979d9ac3b251c4c12c6cf6509e6eb36.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26187">issues.joomla.org/tracker/joomla-cms/26187</a>.</sub>

shindebalu commented:
I have tested this item :red_circle: unsuccessfully on d9fbed64bf3815d877f4ab759d28f966ffd00852<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26187">issues.joomla.org/tracker/joomla-cms/26187</a>.</sub>

Neppert commented:
I have tested this item :red_circle: unsuccessfully on d9fbed64bf3815d877f4ab759d28f966ffd00852<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26187">issues.joomla.org/tracker/joomla-cms/26187</a>.</sub>

webfeuerflo commented:
I have tested this item :white_check_mark: successfully on d9fbed64bf3815d877f4ab759d28f966ffd00852<br /><br />The patch works, but it is not very clear, what this list indicates<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26187">issues.joomla.org/tracker/joomla-cms/26187</a>.</sub>

Quiviro commented:
I have tested this item :white_check_mark: successfully on d9fbed64bf3815d877f4ab759d28f966ffd00852<br /><br />As [Renuka](https://github.com/joomla/joomla-cms/pull/26187#issuecomment-544121842 "") said, it works correctly though I am not sure that way is the intended way<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26187">issues.joomla.org/tracker/joomla-cms/26187</a>.</sub>

brianteeman commented:
Maybe someone can convince the accessibility team to test this

roland-d commented:
This can be closed and picked up in another PR if someone is interested as topic starter is no longer available.
